Source/santad/Logs/SNTSimpleMaildir.m

```diff
@@ -0,0 +1,264 @@
+/// Copyright 2021 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#import "Source/santad/Logs/SNTSimpleMaildir.h"
+
+#include <malloc/malloc.h>
+#include <stdio.h>
+
+#import "Source/common/SNTLogging.h"
+
+/** Helper for creating errors. */
+static NSError *MakeError(NSString *description) {
+  return [NSError errorWithDomain:@"com.google.santa"
+                             code:1
+                         userInfo:@{NSLocalizedDescriptionKey : description}];
+}
+
+size_t SNTRoundUpToNextPage(size_t size) {
+  const size_t pageSize = 4096;
+
+  if (size % pageSize == 0) {
+    return size;
+  }
+  return pageSize * ((int)(size / pageSize) + 1);
+}
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation SNTSimpleMaildir {
```

Comment:
Can you confirm that I understand how this works?

1. Events are written to SNTProtobufEventLog which creates one proto per event and the last thing each of those handlers does is call logEvent: in this class.
2. logEvent: appends each proto to an array in memory as part of a container proto with a single repeated field holding each of the individual events.
3. Every 5s and/or every time the serialized size of the container goes over a size threshold the container proto is written out to a temporary directory and then moved into the real output dir atomically. If the move fails the temp file is removed.
4. After writing, a new container proto is created and the process begins afresh

If that is all correct, I have a few questions:

1. Are the calls to serializedSize not somewhat expensive? My understanding is that call is not cached so the underlying messages are being serialized for each call.
2. Will this not result in many thousands of files being written out?

Reply:
That understanding sounds correct to me! To answer your questions:

1.) Yes, `serializedSize` is definitely expensive. This code was adopted from another project (I'll send you details out-of-band!). I wrote the benchmark utility in part to understand the overall impact of this change as well as help identify benchmarks. Locally, I have modified this file to cache the per-event serialized sizes (plus a bit of overhead needed to account for the array type). I'll have these changes cleaned up and committed before opening the full PR.

2.) Yes, it will result in a bunch of files based on the "file size", "spool size", and "frequency" parameters. Locally I have made these all configurable values and will be committing this change as well for the real PR. The "spool size" param caps the overall amount of space that will be consumed. Once exceeded, no more files will be created. It is assumed that there is some other service draining the completed files to ensure logging doesn't stop.